[CodeGen][ARM64EC] Add support for hybrid_patchable attribute.

cjacek opened the pull request:
This PR implements LLVM part of hybrid_patchable support. A prototype of clang part is here: https://github.com/cjacek/llvm-project/commit/e4b1ffc1d76721af47a7185ace89ec9cc91a280e. The attribute is mentioned in the official documentation: https://learn.microsoft.com/en-us/windows/arm/arm64ec-abi. I described more details about how they work: https://wiki.winehq.org/ARM64ECToolchain#Patchable_functions.

cjacek committed:
[CodeGen][ARM64EC] Add support for hybrid_patchable attribute.

efriedma-quic reviewed (commented):
Maybe also add a tests with varargs and sret to make sure the musttail call correctly preserves all the relevant registers.

cjacek commented on llvm/lib/Target/AArch64/AArch64Arm64ECCallLowering.cpp:
```diff
@@ -670,10 +746,40 @@ bool AArch64Arm64ECCallLowering::runOnModule(Module &Mod) {
+      std::string OrigName(F.getName());
+      F.setName(MangledName.value() + "$hp_target");
+
+      Function *EF = Function::Create(
```
GlobalAlias seems to work fine out of the box for the mangled alias (which points to the thunk). However, global aliases need to point to function definition, while we need to emit a weak alias to an undefined symbol for the unmangled alias. I worked around it by overriding `AsmPrintet::emitGlobalAlias` and special-casing those aliases based on aliasee's metadata. `emitGlobalAlias` was already a virtual function, but I needed to change it to protected in `AsmPrinter` to be able to fall back to it in aarch64 code. This allowed to reduce the gap between IR and generated code, but it's not really eliminated (FWIW, I think it would be possible to similarly define aliases for anti-dependencies if we want to).

efriedma-quic replied:
The current version seems okay.  The whole EXP+ thing would require a significant IR extension to properly represent, which doesn't seem worth it.  (We could theoretically define a Constant specifically to refer to the linker-generated hybrid_patchable thunk of a function, and allow aliases to refer to such a constant, but that seems like overkill if it would only be used in this one place.)

Maybe worth adding a comment in the code describing the intentional gap between a normal alias and what we're emitting here.

Not sure about anti-dep aliases; I still haven't quite wrapped my head around the semantics of antidep.  I guess from the compiler's perspective, it might be close enough to a weak alias that we could just get away with setting the linkage to "weak", and adding some metadata to mark it as anti-dep?

cjacek commented:
The new version fixes a cast in AsmPrinter. It also doesn't use the map to iterate over patchable thunks as it did not guarantee predictable ordering. I also rebased it as CI seemed to have additional unrelated failures.

efriedma-quic commented:
Please add a testcase for taking the address of a hybrid_patchable function.  (Not a complete review; I'll look again next week.)

cjacek commented:
I added tests for taking the address. I also added a test of dllexport. In this case, the output looks weird, but that's compatible with MSVC. I think we could as well not be compatible here and make it work with something like https://github.com/cjacek/llvm-project/commit/4febef20ab27ef4440e9dccf3cc65dd6be1f4d7c on top of this PR.

cjacek commented:
Thanks for the review. The new version uses DenseMap, removes redundant null check and bitcasts and adds a comment explaining the gap.

BTW, for anti-dep aliases, I'm still unsure if I'm missing something, but the only general differences from weak aliases I spot so far is that it's legal to have them defined multiple time and if regular weak alias is present, it overrides anti-deps. Other than that, there is a very specific case where linker seems to use that to detect unmangled -> mangled alias and apply some special-casing.

cjacek commented:
I rebased to resolve the conflict.

efriedma-quic reviewed (approved):
LGTM

(Sorry about the delay; forgot this was still waiting for me.)

nikic commented:
Looks like this attribute is missing specification in LangRef. It also looks like only bitcode writer support was implemented for it, but not bitcode reader support?

cjacek commented:
Good point, thanks. I created #219804 and #219806 for missing parts.
